libraries/standard/mqtt/src/mqtt.c

```diff
@@ -971,7 +1003,10 @@ static MQTTStatus_t receiveConnack( const MQTTContext_t * pContext,
         /* Reading the remainder of the packet by transport recv.
          * Attempt to read once even if the timeout has expired at this point.
          * Invoking receivePacket with remainingTime as 0 would attempt to
-         * recv from network once.*/
+         * recv from network once. If using retries, the remainder of the
+         * CONNACK packet is tried to be read only once. Reading once would be
+         * good as the packet type and remaining length was already read. Hence,
+         * the probability of the remaining 2 bytes available to read is very high. */
         if( pIncomingPacket->type == MQTT_PACKET_TYPE_CONNACK )
         {
             status = receivePacket( pContext,
```

Comment:
If the passed timeout duration has transpired (i.e. `remainingTimeMs` is 0), calling transport recv could add additional delay depending on the platform specific implementation of the recv function. which yields non-deterministic behavior across platforms.
Has this scenario been considered? The extra delay (beyond the total timeout duration) in execution time of the function wouldn't matter if timeout >> transport receive timeout value, however, in the case of the two timeout values being comparable (in scale), the extra delay of the function could be significant (relatively). 

May be it is not a problem, if we have the assumption that 2 extra bytes should have been received in the network buffer if the CONNACK packet type and remaining length were

Reply:
The timeout is not meant to be the maximum time that the function will take, but rather the minimum time (unless there's an error). We document this in the parameter descriptions: https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/development/libraries/standard/mqtt/include/mqtt.h#L259

Reply:
I think Muneeb's comment helped clear the first part of the question. I will take an action on me to update the documentation of MQTT_Connect() as well to match the above mentioned documentation.

For the second part, by the time we find that we could get a valid packet type and remaining length, the possibility is very high that we get the remainder of 2 bytes as well for connack.